src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -143,15 +153,30 @@ public AnalyzerConfigOptionsResult GetOptionsForSourcePath(string sourcePath)
                                 diagnosticBuilder,
                                 config.PathToFile,
                                 _diagnosticIdCache);
+                            sectionBuilder.Add(section);
                         }
                     }
                 }
             }
 
-            return new AnalyzerConfigOptionsResult(
-                treeOptionsBuilder.Count > 0 ? treeOptionsBuilder.ToImmutable() : SyntaxTree.EmptyDiagnosticOptions,
-                analyzerOptionsBuilder.Count > 0 ? analyzerOptionsBuilder.ToImmutable() : AnalyzerConfigOptions.EmptyDictionary,
-                diagnosticBuilder.ToImmutableAndFree());
+            // Try to avoid creating extra dictionaries if we've already seen an options result with the
+            // exact same options
+            var sectionKey = sectionBuilder.ToImmutableAndFree();
```

Comment:
Using an `ImmutableArray<Section>` here means that we are at minimum allocating an array for every lookup attempt. That's even in the good case where we have a cache hit and all the builders return optimal values. Why not instead use `List<Section>` as the key? This is a completely private cache hence it's relatively easy to guard against mutations. That would also mean in the optimal case the lookup is allocation free because we'll get back a `List<Section>` that already has sufficient size to build up the key.